neqo-http3/src/push_controller.rs

```diff
@@ -25,15 +170,324 @@ impl Display for PushController {
+                        stream_id,
+                        events: Vec::new(),
+                    };
+                    Ok(true)
+                }
+                PushState::PushPromise { headers } => {
+                    let tmp = mem::replace(headers, Vec::new());
+                    *push_state = PushState::Active {
+                        stream_id,
+                        headers: tmp,
+                    };
+                    Ok(true)
+                }
+                // The following state have already have a push stream:
+                // PushState::OnlyPushStream | PushState::Active
+                _ => {
+                    qerror!("Duplicate push stream.");
+                    Err(Error::HttpId)
+                }
+            },
+        }
+    }
+
+    fn check_push_id(&mut self, push_id: u64) -> Res<()> {
+        // Check if push id is greater than what we allow.
+        if push_id > self.current_max_push_id {
+            qerror!("Push id is greater than current_max_push_id.");
+            Err(Error::HttpId)
+        } else {
+            Ok(())
+        }
+    }
+
+    pub fn handle_cancel_push(
+        &mut self,
+        push_id: u64,
+        conn: &mut Connection,
+        base_handler: &mut Http3Connection,
+    ) -> Res<()> {
+        qtrace!("CANCEL_PUSH frame has been received, push_id={}", push_id);
+
+        self.check_push_id(push_id)?;
+
+        match self.push_streams.close(push_id) {
+            None => {
+                qtrace!("Push has already been closed (push_id={}).", push_id);
+                Ok(())
+            }
+            Some(ps) => match ps {
+                PushState::Init => Ok(()),
+                PushState::PushPromise { .. } => {
+                    self.conn_events.remove_events_for_push_id(push_id);
+                    self.conn_events.push_canceled(push_id);
+                    Ok(())
+                }
+                PushState::OnlyPushStream { stream_id, .. } => {
+                    qerror!(
+                        "A server should not send CANCEL_PUSH after a push stream has been opened."
+                    );
+                    let _ = base_handler.stream_reset(
+                        conn,
+                        stream_id,
+                        Error::HttpRequestCancelled.code(),
+                    );
+                    Ok(())
+                }
+                PushState::Active { stream_id, .. } => {
+                    qerror!(
+                        "A server should not send CANCEL_PUSH after a push stream has been opened."
+                    );
+                    let _ = base_handler.stream_reset(
+                        conn,
+                        stream_id,
+                        Error::HttpRequestCancelled.code(),
+                    );
+                    self.conn_events.remove_events_for_push_id(push_id);
+                    self.conn_events.push_canceled(push_id);
```

Comment:
Any reason that you would not want to do both of these in OnlyPushStream as well?  Then you could collapse the two arms.